### PR TITLE
CBG-359: Return multiple warnings and move to validate()

### PR DESCRIPTION
**Pull request by JRascagneres**

### CODE REVIEW

**Review by bbrks — CHANGES_REQUESTED** [review]

File: rest/config.go [hunk 1]
```diff
@@ -1222,6 +1228,13 @@ func ServerMain(runMode SyncGatewayRunMode) {
 		logFn()
 	}
 
-	ValidateConfigOrPanic(runMode)
+	// Validation
+	if errorMsgs := config.setupAndValidateDatabases(); errorMsgs != nil && len(errorMsgs) > 0 {
+		for _, err := range errorMsgs {
+			base.Errorf(base.KeyAll, "An error %v", err)
```
Comment (adamcfraser) [inline rest/config.go]: Improve this to "Error during config validation:"

File: rest/config.go [hunk 2]
```diff
@@ -816,18 +839,24 @@ func (config *ServerConfig) deprecatedConfigLoggingFallback() (warnings []base.D
...
 }
 
-func (config *ServerConfig) validateDbConfig(dbConfig *DbConfig) error {
+func (config *ServerConfig) validateDbConfig(dbConfig *DbConfig) []error {
 
 	dbConfig.modifyConfig()
 
 	switch config.RunMode {
 	case SyncGatewayRunModeNormal:
+		if config.HasAnyIndexWriterConfiguredDatabases() {
```
Comment (bbrks) [inline rest/config.go]: Need to re-add `skipRunModeValidation` handling here

**Review by bbrks — CHANGES_REQUESTED** [review]
Looks good - Might just need to re-add the `skipRunModeValidation` handling

File: rest/config.go [hunk 3]
```diff
@@ -816,18 +839,24 @@ func (config *ServerConfig) deprecatedConfigLoggingFallback() (warnings []base.D
...
 	dbConfig.modifyConfig()
 
 	switch config.RunMode {
 	case SyncGatewayRunModeNormal:
+		if config.HasAnyIndexWriterConfiguredDatabases() && !config.SkipRunmodeValidation {
+			base.Panicf(base.KeyAll, "SG is running in normal mode but there are databases configured as index writers")
+		}
 		return dbConfig.validateSgDbConfig()
 	case SyncGatewayRunModeAccel:
+		if config.HasAnyIndexReaderConfiguredDatabases() && !config.SkipRunmodeValidation {
```
Comment (bbrks) [inline rest/config.go]:
```suggestion
		if !config.SkipRunmodeValidation && config.HasAnyIndexReaderConfiguredDatabases() {
```

File: rest/config.go [hunk 4]
```diff
@@ -816,18 +839,24 @@ func (config *ServerConfig) deprecatedConfigLoggingFallback() (warnings []base.D
...
 }
 
-func (config *ServerConfig) validateDbConfig(dbConfig *DbConfig) error {
+func (config *ServerConfig) validateDbConfig(dbConfig *DbConfig) []error {
 
 	dbConfig.modifyConfig()
 
 	switch config.RunMode {
 	case SyncGatewayRunModeNormal:
+		if config.HasAnyIndexWriterConfiguredDatabases() && !config.SkipRunmodeValidation {
```
Comment (bbrks) [inline rest/config.go]:
```suggestion
		if !config.SkipRunmodeValidation && config.HasAnyIndexWriterConfiguredDatabases() {
```

**Review by bbrks — APPROVED** [review]

**Pull request merged**